File: src/c04_professional_classes/open_closed/refactored/OpenClosedPrincipleRefactored.java
package c04_professional_classes.open_closed.refactored;

import java.util.List;

/*
In this example:
- Each shape class (Rectangle, Circle) implements a getArea method from the Shape interface.
- AreaCalculator uses polymorphism to calculate the total area without needing to know the specifics of each shape.
- This design adheres to the Open/Closed Principle since new shapes can be added by extending Shape without modifying AreaCalculator.
*/

// Shape interface enforcing the getArea method
interface Shape {
    double getArea();
}

// Rectangle class implementing Shape
class Rectangle implements Shape {
    private double width;
    private double height;

    public Rectangle(double width, double height) {
        this.width = width;
        this.height = height;
    }

    @Override
    public double getArea() {
        return width * height;
    }
}

// Circle class implementing Shape
class Circle implements Shape {
    private double radius;

    public Circle(double radius) {
        this.radius = radius;
    }

    @Override
    public double getArea() {
        return Math.PI * radius * radius;
    }
}

// AreaCalculator class using polymorphism
class AreaCalculator {
    public double calculateTotalArea(List<Shape> shapes) {
        double totalArea = 0.0;
        for (Shape shape : shapes) {
            totalArea += shape.getArea();
        }
        return totalArea;
    }
}

// Main class to demonstrate the functionality
public class OpenClosedPrincipleRefactored {
    public static void main(String[] args) {
        Rectangle rect = new Rectangle(5, 10);
        Circle circ = new Circle(7);

        List<Shape> shapes = List.of(rect, circ);

        AreaCalculator calculator = new AreaCalculator();
        System.out.println("Total Area: " + calculator.calculateTotalArea(shapes));
    }
}

File: src/c04_professional_classes/matlab/OpenClosedPrincipleRefactored.m
function OpenClosedPrincipleRefactored()
    % In this example:
    % - Each shape class (RectangleRef, CircleRef) implements a getArea method
    %   from the Shape interface (abstract base class).
    % - AreaCalculatorRef uses polymorphism to calculate the total area without
    %   needing to know the specifics of each shape.
    % - This design adheres to the Open/Closed Principle since new shapes can be
    %   added by extending Shape without modifying AreaCalculator.

    rect = RectangleRef(5, 10);
    circ = CircleRef(7);

    shapes = {rect, circ};

    calculator = AreaCalculatorRef();
    fprintf('Total Area: %f\n', calculator.calculateTotalArea(shapes));
end

% Shape interface enforcing the getArea method
classdef Shape < handle
    methods (Abstract)
        area = getArea(obj)
    end
end

% Rectangle class implementing Shape
classdef RectangleRef < Shape
    properties (Access = private)
        width
        height
    end
    methods
        function obj = RectangleRef(width, height)
            obj.width = width;
            obj.height = height;
        end
        function a = getArea(obj)
            a = obj.width * obj.height;
        end
    end
end

% Circle class implementing Shape
classdef CircleRef < Shape
    properties (Access = private)
        radius
    end
    methods
        function obj = CircleRef(radius)
            obj.radius = radius;
        end
        function a = getArea(obj)
            a = pi * obj.radius^2;
        end
    end
end

% AreaCalculator class using polymorphism
classdef AreaCalculatorRef
    methods
        function totalArea = calculateTotalArea(obj, shapes)
            totalArea = 0.0;
            for i = 1:length(shapes)
                shape = shapes{i};
                totalArea = totalArea + shape.getArea();
            end
        end
    end
end
